### DISCUSSION TITLE
Nodes around UI entities

### DISCUSSION THREAD

**Discussion by hydroper** [bevyengine/bevy#9410] [Ideas]
I wonder if it could make sense to have nodes (probably atomically-reference-counted) around UI entities because the existing use of spawning and unspawning commands supports no children operations, such as finding child by path (whether `../child_id/blah`), getting a child's path. Consider:

- Godot nodes
  - Node paths are similiar to Unix paths, consisting of slashes, dots, `..` and names
- Angular's components
  - Have their template's host element as an ordinary custom HTML element.
  - Can manipulate elements contained within the template (such as `@ViewModel(...) elementRef: ElementRef;` and `elementRef.nativeElement`).

I think [this is a way](https://github.com/rialight/api#graphics-markup-and-custom-nodes) to implement nodes in Rust with composition instead of traits for avoiding dynamic dispatch on complementary fields:

Summary:

- `Node` avoids contravariant methods. For example, if `set_foo` belongs to `Foo` variant, generally `Node` doesn't define it.
- `Node` implements `From<K>` (so a contravariant type can convert into `Node` by simply `.into()`).
- 3 macros are defined using the `syn` crate: `markup!`, `define_node!` and `define_ui_component!`.
- `define_node!` defines a node kind, which holds (_node_, _data_), inheriting everything from _node_, implicitly 
defining `set_` methods for attributes used in the `markup!` macro. _data_ is of the hidden `K__Internal` type.
- `to::<K>`, `try_into::<K>` and `.is::<K>` are supported by `Node`.

**Comment by hydroper** [comment]
I think this doesn't fit with Bevy after looking at `bevy::hierarchy`. I think there are no "entity paths" in this sense, but I'm not sure it's necessary

**Reply by nicopap** [reply]
Bevy has "entity paths". It's used in `bevy_animation`. Simply add a `Name` component to the entities in the path. The path components are the name of each entity in the hierarchy. But nodes as you describe it have little to do with ECS. You seem to describe an OOP game system. Which is incompatible with rust, as rust has no concept of inheritance (and for good reasons!). It's not impossible to create fake inheritance in rust (as your link demonstrates), but it is complex and you won't gain anything from it. I invite you to approach bevy without preconceptions. It helps to think of Rust as a non-OOP language.

**Reply by hydroper** [reply]
@nicopap Interesting, but I've not found any way to get a child entity by a path anywhere
